pallets/foreign-investments/src/types.rs

```diff
@@ -0,0 +1,379 @@
+// Copyright 2021 Centrifuge Foundation (centrifuge.io).
+// This file is part of Centrifuge Chain project.
+
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+use cfg_traits::SimpleCurrencyConversion;
+use cfg_types::investments::Swap;
+use codec::{Decode, Encode, MaxEncodedLen};
+use frame_support::{dispatch::fmt::Debug, RuntimeDebugNoBound};
+use scale_info::TypeInfo;
+use sp_runtime::traits::{EnsureAdd, EnsureSub, Zero};
+
+use crate::Config;
+
+/// Reflects the reason for the last token swap update such that it can be
+/// updated accordingly if the last and current reason mismatch.
+#[derive(
+	Clone, Copy, PartialOrd, Ord, PartialEq, Eq, Debug, Encode, Decode, TypeInfo, MaxEncodedLen,
+)]
+pub enum TokenSwapReason {
+	Investment,
+	Redemption,
+}
+
+// TODO: Docs
+pub trait InvestStateConfig {
+	type Balance: Clone + Copy + EnsureAdd + EnsureSub + Ord + Debug + Zero;
+	type CurrencyId: Clone + Copy + PartialEq + Debug;
+	type CurrencyConverter: SimpleCurrencyConversion<
+		Balance = Self::Balance,
+		Currency = Self::CurrencyId,
+		Error = sp_runtime::DispatchError,
+	>;
+}
+
+impl<T: Config> InvestStateConfig for T {
+	type Balance = T::Balance;
+	type CurrencyConverter = T::CurrencyConverter;
+	type CurrencyId = T::CurrencyId;
+}
+
+/// Reflects all states a foreign investment can have until it is processed as
+/// an investment via `<T as Config>::Investment`. This includes swapping it
+/// into a pool currency or back, if the investment is decreased before it is
+/// fully processed.
+#[derive(
+	PartialOrd, Ord, PartialEq, Eq, RuntimeDebugNoBound, Encode, Decode, TypeInfo, MaxEncodedLen,
+)]
+#[scale_info(skip_type_params(T))]
+pub enum InvestState<T: InvestStateConfig> {
+	/// Default state for initialization which will never be actively put into
+	/// chain state, i.e. if this state is the result of applying transition(s),
+	/// then the corresponding `InvestmentState` will be cleared.
+	NoState,
+	/// The investment is waiting to be processed.
+	InvestmentOngoing { invest_amount: T::Balance },
+	/// The investment is currently swapped into the required pool currency.
+	ActiveSwapIntoPoolCurrency {
+		swap: Swap<T::Balance, T::CurrencyId>,
+	},
+	/// The unprocessed investment was fully decreased and is currently swapped
+	/// back into the corresponding foreign currency.
+	ActiveSwapIntoForeignCurrency {
+		swap: Swap<T::Balance, T::CurrencyId>,
+	},
```

Comment:
Is this state distinction really necessary? Could the `Pool/Foreign` information be extracted from the direction of the swap?

Reply:
I honestly think removing this state is not advantageous if we don't refactor the entire `InvestState`s and `RedeemState`s. IMO, it would worsen readability when handling transitions as we would still need to determine the swap direction and whether that works with the given transition. As an example, when we decrease, we error out on `ActiveSwapIntoForeignCurrency`.  This fits into the [initial conversation about refactoring those states](https://github.com/centrifuge/centrifuge-chain/pull/1473#discussion_r1297137596) two weeks ago. Given the timeline has just worsened since then, time is missing to touch that. Again: The current approach is not the most elegant one but it is straight forward and can be reproduced by the diagrams in the spec.

Reply:
Ok, thanks William! It's not a problem at all. Yeah, just was wondering if it added more info to the swap. And it's true that right now, in order to determine the direction it's easier having those split.